### PR TITLE
[E2E] PoC: Run all `question` repros together

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### Status
READY

### What does this PR accomplish?
- Groups all individual reproductions in the `question` e2e group,
- wraps them in a function,
- exports that function
- imports every single individual repro in an aggregate file called `question-reproductions.cy.spec.js` and
- runs all of those repros together

### Why this change?
As I mentioned before, the overhead of tearing down one spec and starting to run another one is 7-10s on average. A lot of these reproductions finish running in under 2-3s. The overhead alone can account for 2-3 minutes of the wasted time.

By grouping all repros together, that spec now needs cca 50s to finish, which is much better for the overall speed.

In my Metabase fork which runs on GitHub basic runners (ubuntu-latest) I'm seeing 5-6 minutes faster runs. That's a significant improvement!
[Run without optimization](https://github.com/nemanjaglumac/metabase/runs/5823557737?check_suite_focus=true)
[Run all repros together](https://github.com/nemanjaglumac/metabase/runs/5824089923?check_suite_focus=true)

| Machine| Job                       | Before | After | Gain  |
|--|---------------------------|--------|-------|-------|
|`ubuntu-latest`|                       |  |  |   |
|| question-oss                  | 23:24   | 18:37  | 4:47  |
|| question-ee                 | 23:35   | 16:33  | 7:02  |
|`buildjet-4vcpu`|                       | | |  |
|| question-oss                  | 12:20   | 10:41  | 1:39  |
|| question-ee                 | 14:36   | 10:29  | 4:07  |

## Note
The downside of this approach is that it takes some time getting used to a new flow of writing a spec in one file and then importing and running it in another file, but overall it shouldn't be too difficult to get used to it.

### CODE REVIEW

File: frontend/test/metabase/scenarios/question/reproductions/13097-mongo-apply-distinct-count-multiple-columns.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { restore, visualize } from "__support__/e2e/cypress";
+
+const MONGO_DB_NAME = "QA Mongo4";
+
+export function issue13097() {
+  describe.skip("issue 13097", () => {
```
Comment (diogormendes) [inline frontend/test/metabase/scenarios/question/reproductions/13097-mongo-apply-distinct-count-multiple-columns.js:6]: Should we still be skipping?

Comment (diogormendes) [inline frontend/test/metabase/scenarios/question/reproductions/13097-mongo-apply-distinct-count-multiple-columns.js:6]: Should we still be skipping?

File: frontend/test/metabase/scenarios/question/reproductions/14957-unable-to-save-question-before-query-executed.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { restore, modal } from "__support__/e2e/cypress";
+
+const PG_DB_NAME = "QA Postgres12";
+
+export function issue14957() {
+  describe.skip("issue 14957", () => {
```
Comment (diogormendes) [inline frontend/test/metabase/scenarios/question/reproductions/14957-unable-to-save-question-before-query-executed.js:6]: Should we still be skipping?

Comment (diogormendes) [inline frontend/test/metabase/scenarios/question/reproductions/14957-unable-to-save-question-before-query-executed.js:6]: Should we still be skipping?

File: frontend/test/metabase/scenarios/question/reproductions/13097-mongo-apply-distinct-count-multiple-columns.js (hunk 1 above)
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/reproductions/13097-mongo-apply-distinct-count-multiple-columns.js:6]: If the issue hasn't been fixed - we should. This doesn't change that functionality.

Now, regarding this specific issue, it seems that Cam fixed it long time ago. I'll have to check was there a specific reason why the associated repro hasn't been unskipped, or did we simply forget to do it.

Comment (diogormendes) [reply frontend/test/metabase/scenarios/question/reproductions/13097-mongo-apply-distinct-count-multiple-columns.js:6]: https://github.com/metabase/metabase/pull/21477

**Review by flamber — COMMENTED** [review]
Wouldn't be easier just to put all tests into a single file instead? This feels like a hack and diverts our implementation further from the "standard".
If not, then there definitely need to be some automation that ensure it generates everything correctly.

Think the approve should come from whoever writes the tests. If everybody is onboard, then I'm good too.

**Review by gusaiani — COMMENTED** [review]
Imo this approach is one step better than what we have, and very meaningful because it speeds up runs so much.

Having all of these tests in one gigantic file, again from my experience and then opinion, adds huge cognitive friction that may compare with, or even surpass, the friction we currently experience from the test run times.

Perhaps it will be possible to automatically loop files in the `./reproductions/` dir, importing them and running the tests. 
In other words, we won't have to manually write the imports and calling them as functions.

I haven't tested this approach myself, but may be worth a try.

https://javascript.tutorialink.com/import-and-execute-all-files-on-folder-with-es6/

**Comment by nemanjaglumac** [comment]
> Wouldn't be easier just to put all tests into a single file instead?

It would, but as @gusaiani said, it would be a nightmare to review such a file and to get around scrolling through 1000+ LoC.

> Having all of these tests in one gigantic file, again from my experience and then opinion, adds huge cognitive friction that may compare with, or even surpass, the friction we currently experience from the test run times.


> This feels like a hack and diverts our implementation further from the "standard".

I wouldn't call it hack since [JavaScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) are a part of the standard and were made to solve exactly these kinds of problems. How to keep everything neat and organized (modular), and not worry about the final compiled JS file that will be executed.

As far as I'm concerned, we don't have to go down this route and I'll gladly let other people who write tests voice their opinions. But choices for me are: 
1. either we don't go with this at all or
2. we find whichever system to keep things modular

The only thing I'm firmly against is keeping everything in one gigantic file.

**Review by gusaiani — APPROVED** [review]
I say let's try it this way, reverting won't be too much work and you can always overrule me.

There's a merge conflict atm, shouldn't be hard to update.

**Comment by nemanjaglumac** [comment]
> There's a merge conflict atm, shouldn't be hard to update.

Yup, I am aware. I caused it by merging https://github.com/metabase/metabase/pull/21477

Let's also wait for @ariya's input on this before we go further. We're in no rush to merge it, anyway.

**Review by ariya — APPROVED** [review]
Let's ship this :rocket:
